apps/fipsinstall.c

```diff
@@ -0,0 +1,397 @@
+
+typedef enum OPTION_choice {
+    OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+    OPT_IN, OPT_CFG, OPT_MACOPT, OPT_VERIFY, OPT_MAC, OPT_SECTION_NAME
+} OPTION_CHOICE;
+
+const OPTIONS fipsinstall_options[] = {
+    {"help", OPT_HELP, '-', "Display this summary"},
+    { OPT_MORE_STR, 0, 0, "e.g: openssl fipsinstall -in fips -cfg "
+      "fips.conf -mac HMAC -macopt digest:SHA256 -macopt hexkey:00"},
+    {"in", OPT_IN, 's', "Input module file name (with no extension)"},
+    {"cfg", OPT_CFG, '>', "config file name to output or verify"},
+    {"mac", OPT_MAC, 's', "MAC name"},
+    {"macopt", OPT_MACOPT, 's', "MAC algorithm control parameters in n:v form. "
+                                "See 'Supported Controls' in the EVP_MAC_ docs"},
+    {"verify", OPT_VERIFY, '-', "Verify the installed status is correct"},
+    {"section", OPT_SECTION_NAME, 's', "the section name in the config file"},
+
+    {NULL}
+};
+
+static int mac_ctrl_string(EVP_MAC_CTX *ctx, const char *value)
+{
+    int rv;
+    char *stmp, *vtmp = NULL;
+
+    stmp = OPENSSL_strdup(value);
+    if (stmp == NULL)
+        return -1;
+    vtmp = strchr(stmp, ':');
+    if (vtmp != NULL) {
+        *vtmp = 0;
+        vtmp++;
+    }
+    rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+    OPENSSL_free(stmp);
+    return rv;
+}
+
+static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
+                  unsigned char *out, size_t *out_len)
+{
+    int ret = 0;
+    int i;
+
+    if (!EVP_MAC_init(ctx))
+        goto err;
+    if (EVP_MAC_size(ctx) > *out_len)
+        goto end;
+    for (;;) {
+        i = BIO_read(in, (char *)tmp, BUFSIZE);
+        if (i < 0)
+            goto err;
+        if (i == 0)
+            break;
+        if (!EVP_MAC_update(ctx, tmp, i))
+            goto err;
+    }
+end:
+    if (!EVP_MAC_final(ctx, out, out_len))
+        goto err;
+    ret = 1;
+err:
+    return ret;
+}
+
+static int hexstr2bio(BIO *out, const unsigned char *data, size_t len)
+{
+    size_t i;
+
+    if (len == 0)
+        return 0;
+
+    for (i = 0; i < len; i++) {
+        BIO_printf(out, "%02X", data[i]);
+    }
+    return 1;
+}
+
+static int print_mac(BIO *bio, const char *label, const unsigned char *mac,
+                     size_t len)
+{
+    if (!BIO_printf(bio, "%s =", label))
+        return 0;
+    if (!hexstr2bio(bio, mac, len))
+        return 0;
+    return BIO_printf(bio, "\n");
+}
+
+static int load_fips_prov_and_run_self_test(char *fname,
+                                            const unsigned char *mac,
+                                            size_t mac_len)
+{
+    char buffer[1024], *buf;
+    size_t sz;
+    int ret = 0;
+    OSSL_PROVIDER *prov = NULL;
+    char *prov_name, *dot;
+    static OSSL_PARAM params[2];
+    OSSL_PARAM param_end = OSSL_PARAM_END;
+    BIO *mem_bio = NULL;
+
+    mem_bio = BIO_new_mem_buf(buffer, sizeof(buffer));
+    if (mem_bio == NULL)
+        return 0;
+
+    if (!hexstr2bio(mem_bio, mac, mac_len))
+        goto end;
```

Comment:
May I suggest using [`OPENSSL_buf2hexstring`](https://www.openssl.org/docs/manmaster/man3/OPENSSL_buf2hexstr.html) instead of reinventing it?